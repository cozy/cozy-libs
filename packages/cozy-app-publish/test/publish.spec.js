/* eslint-env jest */
const fetch = require('jest-fetch-mock')

const publish = require('../lib/publish')

function getOptions() {
  const options = {
    registryEditor: 'cozy',
    registryToken: 'registryTokenForTest123',
    appSlug: 'mock-app',
    appBuildUrl: 'https://mock.getarchive.cc/12345.tar.gz',
    appVersion: '2.1.8-dev.12345',
    registryUrl: 'https://mock.registry.cc',
    spaceName: 'mock_space',
    appType: 'webapp',
    sha256Sum:
      'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
  }
  return options
}

function getObjectToSnapshot(fetchMock, callNumber = 0) {
  return {
    fetchURL: fetchMock.mock.calls[callNumber][0],
    options: fetchMock.mock.calls[callNumber][1]
  }
}

describe('Publish script (helper)', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should work correctly if expected options provided', async () => {
    fetch.mockResponseOnce('', {
      status: 201
    })
    await publish(getOptions())
    expect(fetch).toHaveBeenCalledTimes(1)
    expect(getObjectToSnapshot(fetch)).toMatchSnapshot()
  })

  it('should work correctly if no space name provided', async () => {
    fetch.mockResponseOnce('', {
      status: 201
    })
    const options = getOptions()
    delete options.spaceName
    await publish(options)
    expect(fetch).toHaveBeenCalledTimes(1)
    expect(getObjectToSnapshot(fetch)).toMatchSnapshot()
  })

  it('should handle error message if the publishing failed with 404', async () => {
    fetch.mockResponseOnce(
      JSON.stringify({ error: 'Application slug not found' }),
      {
        status: 404,
        statusText: '(TEST) Not Found'
      }
    )

    expect(publish(getOptions())).rejects.toThrowErrorMatchingSnapshot()
    expect(fetch).toHaveBeenCalledTimes(1)
  })

  it('should handle error message if the publishing failed with an unexpected fetch error', async () => {
    fetch.mockRejectOnce(new Error('(TEST) Unexpected error'))
    expect(publish(getOptions())).rejects.toThrowErrorMatchingSnapshot()
    expect(fetch).toHaveBeenCalledTimes(1)
  })

  describe('when provided space if the default one', () => {
    it('should not put the space name in the requested URL', async () => {
      fetch.mockResponseOnce('', {
        status: 201
      })

      const options = getOptions()
      options.spaceName = 'default'

      await publish(options)

      expect(fetch.mock.calls[0][0]).toBe(
        options.registryUrl + '/registry/' + options.appSlug
      )
    })
  })
})
