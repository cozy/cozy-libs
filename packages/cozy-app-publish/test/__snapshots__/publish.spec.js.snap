// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Publish script (helper) should handle error message if the publishing failed with 404 1`] = `"{\\"error\\":\\"Application slug not found\\"}"`;

exports[`Publish script (helper) should handle error message if the publishing failed with an unexpected fetch error 1`] = `"(TEST) Unexpected error"`;

exports[`Publish script (helper) should work correctly if expected options provided 1`] = `
Object {
  "fetchURL": "https://mock.registry.cc/mock_space/registry/mock-app",
  "options": Object {
    "body": "{\\"editor\\":\\"cozy\\",\\"version\\":\\"2.1.8-dev.12345\\",\\"url\\":\\"https://mock.getarchive.cc/12345.tar.gz\\",\\"sha256\\":\\"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\\",\\"type\\":\\"webapp\\"}",
    "headers": Object {
      "Authorization": "Token registryTokenForTest123",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
}
`;

exports[`Publish script (helper) should work correctly if no space name provided 1`] = `
Object {
  "fetchURL": "https://mock.registry.cc/registry/mock-app",
  "options": Object {
    "body": "{\\"editor\\":\\"cozy\\",\\"version\\":\\"2.1.8-dev.12345\\",\\"url\\":\\"https://mock.getarchive.cc/12345.tar.gz\\",\\"sha256\\":\\"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\\",\\"type\\":\\"webapp\\"}",
    "headers": Object {
      "Authorization": "Token registryTokenForTest123",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
}
`;
